includes/core/server/class-request.php

```diff
@@ -0,0 +1,523 @@
+	 *
+	 * @var bool
+	 */
+	private $protected_mode = false;
+
+	/**
+	 * Holds the API client of WCPay.
+	 *
+	 * @var WC_Payments_API_Client
+	 */
+	protected $api_client;
+
+	/**
+	 * Holds the HTTP interface of WCPay.
+	 *
+	 * @var WC_Payments_Http_Interface
+	 */
+	protected $http_interface;
+
+	/**
+	 * Prevents the class from being constructed directly.
+	 *
+	 * @param WC_Payments_API_Client     $api_client     The API client to use to send requests.
+	 * @param WC_Payments_Http_Interface $http_interface The HTTP interface for the server.
+	 */
+	public function __construct( WC_Payments_API_Client $api_client, WC_Payments_Http_Interface $http_interface ) {
+		$this->api_client     = $api_client;
+		$this->http_interface = $http_interface;
+	}
+
+	/**
+	 * Returns the needed API.
+	 *
+	 * @return string Check WCPay\Core\Server\APIs.
+	 */
+	abstract public function get_api(): string;
+
+	/**
+	 * Returns the method of the request.
+	 *
+	 * @return string See the constants in WordPress's `Requests` class.
+	 */
+	abstract public function get_method(): string;
+
+	/**
+	 * This is a legacy method, and is the same throughout the codebase.
+	 * Might be worth removing while refactoring to use the Core\Server API.
+	 *
+	 * @return bool
+	 */
+	public function is_site_specific(): bool {
+		return true;
+	}
+
+	/**
+	 * If true, the request will be signed with the user token rather than blog token. Defaults to false.
+	 *
+	 * @return bool
+	 */
+	public function should_use_user_token(): bool {
+		return false;
+	}
+
+	/**
+	 * Indicates if the raw response should be returned.
+	 *
+	 * @return bool
+	 */
+	public function should_return_raw_response(): bool {
+		return false;
+	}
+
+	/**
+	 * Returns all of the parameters for the request.
+	 *
+	 * @return array
+	 * @throws Invalid_Request_Parameter_Exception If the request has not been initialized yet.
+	 */
+	final public function get_params() {
+		$defaults = $this->get_default_params();
+		$params   = array_merge( $defaults, $this->params );
+
+		$missing_params = [];
+		foreach ( $this->get_required_params() as $name ) {
+			if ( ! isset( $params[ $name ] ) ) {
+				$missing_params[] = $name;
+			}
+		}
+
+		if ( ! empty( $missing_params ) ) {
+			throw new Invalid_Request_Parameter_Exception(
+				sprintf(
+					'Trying to access the parameters of a request which is not (fully) initialized yet. Missing parameter(s) for %s: %s',
+					get_class( $this ),
+					implode( ', ', $missing_params )
+				),
+				'wcpay_core_invalid_request_parameter_missing_parameters'
+			);
+		}
+
+		foreach ( $params as $key => $value ) {
+			if ( true === $value ) {
+				// The WCPay server requires the string 'true'.
+				$params[ $key ] = 'true';
+			}
+		}
+
+		return $params;
+	}
+
+	/**
+	 * Get request param by key.
+	 *
+	 * @param string $key Key to get.
+	 *
+	 * @return mixed
+	 * @throws Invalid_Request_Parameter_Exception
+	 */
+	final public function get_param( $key ) {
```

Comment:
I did not add it in the beginning, because within the gateway it should not have a case, but consumers might benefit from checking the existing parameters of the request, so I prefer to keep it.